### PR TITLE
Compress PT2E in Executorch

### PR DESCRIPTION
**Pull request by anzr299**
### Summary
[PLEASE REMOVE] See [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests) for ExecuTorch PR guidelines.

[PLEASE REMOVE] If this PR closes an issue, please add a `Fixes #<issue-id>` line.

[PLEASE REMOVE] If this PR introduces a fix or feature that should be the upcoming release notes, please add a "Release notes: <area>" label. For a list of available release notes labels, check out [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests).

### Test plan
[PLEASE REMOVE] How did you test this PR? Please write down any manual commands you used and note down tests that you have written if applicable.

### CODE REVIEW

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]

File: backends/openvino/requirements.txt [hunk 1]
```diff
@@ -1,2 +1,2 @@
 transformers
-git+https://github.com/openvinotoolkit/nncf@3d753ac#egg=nncf
+git+https://github.com/openvinotoolkit/nncf#egg=nncf
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/requirements.txt]: Yes, that's a good point. I forgot to make it fixed since I assumed it was just develop branch. I will point it to a fixed commit.

File: backends/openvino/quantization/nncf_compression.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            if pos >= len(token_list):
+                token_list.append(torch.argmax(logits[:], dim=-1).item())
+    token_list = [
+        (
+            pos,
+            token,
+        )
+        for pos, token in enumerate(token_list)
+    ]
+    return token_list
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantization/nncf_compression.py]: From executorch quantization flow's calibration method: https://github.com/anzr299/executorch/blob/535a19898ee829d939a9d8c762cb89dae9231dd7/extension/llm/export/builder.py#L289

File: examples/models/llama/export_llama_lib.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 from typing import Callable, List, Optional, Union
 
 import torch
+from executorch.backends.openvino.quantization import apply_nncf_data_aware_compression
```
[the remark this answers is not in the document]
Comment (anzr299) [reply examples/models/llama/export_llama_lib.py]: All other backends directly specify the intiialization in https://github.com/anzr299/executorch/blob/535a19898ee829d939a9d8c762cb89dae9231dd7/examples/models/llama/export_llama_lib.py#L774. 

My intitial approach was to keep the function `apply_nncf_data_aware_compression` in the `export_llama_lib.py` file itself, but mustafa suggested to move it to a seperate file which will sit in our backend directory `executorch/backends/llama` This was as per the suggestions of the executorch team to mustafa. 

If we talk about something similar, where backends like qualcomm have some completely unique implementations for lets say attention, batching and so on. They have created a completely new example and files and stored it all stored in their backend directory as well.

File: examples/models/llama/export_llama_lib.py [hunk 4]
```diff
@@ -920,9 +936,14 @@ def _to_edge_and_lower_llama_openvino(
     for partitioner in partitioners:
         logging.info(f"--> {partitioner.__class__.__name__}")
 
-    builder = builder_exported.pt2e_quantize(quantizers).to_edge_transform_and_lower(
-        partitioners
-    )
+    if awq or scale_estimation:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply examples/models/llama/export_llama_lib.py]: I suggest to always use `compress_pt2e`, as we are advertising it as an advanced torch.ao quantization

File: backends/openvino/quantizer/quantizer.py [hunk 5]
```diff
@@ -119,15 +126,18 @@ def __init__(
...
                 **kwargs,
             )
             subset_size = 1  # Doesn't really matter in this case since it is data-free. Should just be +ve
-            self._algo = nncf.quantization.algorithms.weight_compression.algorithm.WeightCompression(
-                subset_size=subset_size, **weight_compression_configuration
+            self._algo = WeightCompression(
+                subset_size=subset_size, **self._wc_config
             )
 
+    def get_wc_algo_configuration(self):
```
Comment (anzr299) [inline backends/openvino/quantizer/quantizer.py]: This method is used by the adapter in the PR: https://github.com/openvinotoolkit/nncf/pull/3782

File: backends/openvino/quantizer/quantizer.py [hunk 6]
```diff
@@ -124,7 +130,7 @@ def __init__(
                 **kwargs,
             )
             subset_size = 1  # Doesn't really matter in this case since it is data-free. Should just be +ve
-            self._algo = nncf.quantization.algorithms.weight_compression.algorithm.WeightCompression(
+            self._algo = WeightCompression(
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: Done

File: backends/openvino/quantizer/quantizer.py [hunk 7]
```diff
@@ -163,6 +169,16 @@ def get_nncf_quantization_setup(
         self._algo._set_backend_entity(model)
         return self._algo.find_quantization_setup(model, nncf_graph)
 
+    def get_nncf_weight_compression_parameters(
+        self,
+        model: torch.fx.GraphModule,
+        nncf_graph: NNCFGraph,
+    ) -> tuple[
+        list[WeightCompressionParameters], Optional[dict[str, WCTensorStatistic]]
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: You're right. Fixed

File: backends/openvino/quantizer/quantizer.py [hunk 8]
```diff
@@ -163,6 +169,16 @@ def get_nncf_quantization_setup(
         self._algo._set_backend_entity(model)
         return self._algo.find_quantization_setup(model, nncf_graph)
 
+    def get_nncf_weight_compression_parameters(
+        self,
+        model: torch.fx.GraphModule,
+        nncf_graph: NNCFGraph,
+    ) -> tuple[
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py:195]: This method is used by the openvino adapter

File: backends/openvino/requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/requirements.txt:2]: I will add the commit after https://github.com/openvinotoolkit/nncf/pull/3782/files is merged

File: examples/models/llama/export_llama_lib.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anzr299) [reply examples/models/llama/export_llama_lib.py]: As per our discussion offline, I have made it a lazy import here: https://github.com/anzr299/executorch/blob/d761c3be9c809a78690633393cf0df6e67088cd3/examples/models/llama/export_llama_lib.py#L939

File: examples/models/llama/export_llama_lib.py [hunk 9]
```diff
@@ -241,6 +242,19 @@ def build_args_parser() -> argparse.ArgumentParser:
...
+        action="store_true",
+        help="Whether to use AWQ from NNCF. Applicable only for the Openvino backend.",
+    )
+
+    parser.add_argument(
+        "--nncf_scale_estimation",
+        action="store_true",
+        help="Whether to use Scale Estimation algorithm from NNCF. Applicable only for the Openvino backend",
+    )
+
```
[the remark this answers is not in the document]
Comment (anzr299) [reply examples/models/llama/export_llama_lib.py:256]: Done

File: backends/openvino/quantization/__init__.py [hunk 10]
```diff
@@ -0,0 +1,3 @@
+from .nncf_compression import use_nncf_compression
+
+__all__ = ["use_nncf_compression"]
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantization/__init__.py]: This is to import the function that applies nncf compress pt2e. The name was wrong here. Fixed it

File: backends/openvino/quantizer/llm_compression.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+            if pos >= len(token_list):
+                token_list.append(torch.argmax(logits[:], dim=-1).item())
+    token_list = [
+        (
+            pos,
+            token,
+        )
+        for pos, token in enumerate(token_list)
+    ]
+    return token_list
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantizer/llm_compression.py:51]: Done

File: backends/openvino/quantization/nncf_compression.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def transform_fn(token_pos_map: tuple[int, str]):
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantization/nncf_compression.py]: Done

File: backends/openvino/quantization/nncf_compression.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    # tokenized_text = tokenizer.encode(prompts, bos=False, eos=False)
+    inputs = ()
```
[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantization/nncf_compression.py]: Done

File: examples/models/llama/export_llama_lib.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anzr299) [reply examples/models/llama/export_llama_lib.py]: Done

File: backends/openvino/quantizer/llm_compression.py [hunk 14]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Intel Corporation
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply backends/openvino/quantizer/llm_compression.py:1]: You sure this llama example specific code is belong to this directory? I'm not sure. How other backends are doing this?

[the remark this answers is not in the document]
Comment (anzr299) [reply backends/openvino/quantizer/llm_compression.py:1]: Hm, perhaps I can keep get_calibration_data and move the code from apply_nncf_data_aware_compression and and move the to the example. 

Although, these methods can be used for any LLM which is an object of LLMEdgeManager. This is because get_calibration_data is a common function in executorch and transform_func does not perform any special task

File: backends/openvino/quantizer/quantizer.py [hunk 15]
```diff
@@ -119,15 +123,18 @@ def __init__(
             compression_mode = mode.value.replace(
                 "wo", ""
             )  # Mode value has to match NNCF CompressWeightsMode
-            weight_compression_configuration = get_weight_compression_configuration(
+            self._wc_config = get_weight_compression_configuration(
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/quantizer.py]: Let's do `algo.get_config()` and sort outputs to keep only required params there, then we don't have to keep copies in `self._wc_config`

File: backends/openvino/quantizer/quantizer.py [hunk 16]
```diff
@@ -119,15 +123,18 @@ def __init__(
             compression_mode = mode.value.replace(
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/quantizer.py]: Please use dict there

File: backends/openvino/quantizer/llm_compression.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply backends/openvino/quantizer/llm_compression.py:1]: Let's move this file to the `quantizer/llm_compression.py`

File: backends/openvino/quantization/nncf_compression.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    inputs = (
+        torch.tensor(token_pos_map[1]).unsqueeze(0).unsqueeze(0),
+        {"input_pos": torch.tensor([token_pos_map[0]])},
+    )
+
+    return inputs
+
+
+def apply_nncf_data_aware_compression(
+    builder_exported, quantizers, awq: bool, scale_estimation: bool
```
Comment (daniil-lyakhov) [inline backends/openvino/quantization/nncf_compression.py]: typehints

File: backends/openvino/quantizer/quantizer.py [hunk 18]
```diff
@@ -163,6 +170,18 @@ def get_nncf_quantization_setup(
         self._algo._set_backend_entity(model)
         return self._algo.find_quantization_setup(model, nncf_graph)
 
+    def get_nncf_weight_compression_parameters(
+        self,
+        model: torch.fx.GraphModule,
+        nncf_graph: NNCFGraph,
+    ) -> tuple[
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/quantizer.py:195]: Description please

File: backends/openvino/quantizer/quantizer.py [hunk 19]
```diff
@@ -119,15 +123,18 @@ def __init__(
...
                 nncf.CompressWeightsMode(compression_mode),
                 **kwargs,
             )
             subset_size = 1  # Doesn't really matter in this case since it is data-free. Should just be +ve
             self._algo = nncf.quantization.algorithms.weight_compression.algorithm.WeightCompression(
-                subset_size=subset_size, **weight_compression_configuration
+                subset_size=subset_size, **self._wc_config
             )
 
+    def get_wc_algo_configuration(self):
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/quantizer.py]: Description

File: backends/openvino/quantizer/quantizer.py [hunk 20]
```diff
@@ -119,15 +123,18 @@ def __init__(
...
+        return self._wc_config
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/quantizer.py]: General comment: try to avoid returning mutable objects as they are, it is always better to return a copy of a dict/list because it saves code from unexpected data modifications

Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: Alright. The config is processed here and only the selected configs which we need are repacked into a dict and returned.

File: backends/openvino/quantizer/quantizer.py (hunk 19 above)
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: Done

File: backends/openvino/quantizer/quantizer.py (hunk 18 above)
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py:195]: Done

File: backends/openvino/quantization/nncf_compression.py (hunk 17 above)
Comment (anzr299) [reply backends/openvino/quantization/nncf_compression.py]: Done

File: backends/openvino/quantizer/quantizer.py (hunk 16 above)
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: Done

File: backends/openvino/quantizer/quantizer.py (hunk 15 above)
Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: Done

File: backends/openvino/quantizer/llm_compression.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+
+# This code is taken from https://github.com/pytorch/executorch/blob/0c54fd0483314da173f8e14d63d2ed9591c7133a/extension/llm/export/builder.py#L278
+def get_calibration_data(
+    module: torch.fx.GraphModule, tokenizer, prompts: str, max_len: int
+):
+    """
+    This method is used to obtain calibration data from a prompt so that the algorithm
+    is calibrated not only with the dataset but also the inputs which are output-ed by
+    the model.
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/llm_compression.py:28]: Please specify a model/ models for which this method is applicable

File: backends/openvino/quantizer/llm_compression.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            token,
+        )
+        for pos, token in enumerate(token_list)
+    ]
+    return token_list
+
+
+def transform_fn(token_pos_map: tuple[int, int]):
+    """
+    Transforms and returns input from dataset so that it is acceptable by the model
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/llm_compression.py:56]: Please specify model/models for which this transform model fn is working

File: backends/openvino/quantizer/quantizer.py [hunk 23]
```diff
@@ -116,18 +120,35 @@ def __init__(
...
+            weight_compression_configuration["subset_size"] = 1  # Doesn't really matter in this case since it is data-free. Should just be +ve
             self._algo = nncf.quantization.algorithms.weight_compression.algorithm.WeightCompression(
-                subset_size=subset_size, **weight_compression_configuration
+                **weight_compression_configuration
             )
 
+    def get_quantizer_specific_internal_algo_config(self) -> Dict[str, Any]:
+        """
+        Returns the config attributes like all_layers, group_size, backup_mode and Quantization mode
+        from the quantizer's internal algorithm. These attributes are defined when initializing the quantizer
```
Comment (daniil-lyakhov) [inline backends/openvino/quantizer/quantizer.py]: It is not yet a data class which would specify the interface for compress_pt2e, but let's describe all the params in the output dict. I believe this would help on the code review
```suggestion
        Returns a dictionary with all_layers, group_size, backup_mode and Quantization mode parameters
        used by the compress_pt2e weight compression algorithm. 
        * all_layers - desc...
        * group_size - desc..
        ...
        ```

Comment (anzr299) [reply backends/openvino/quantizer/quantizer.py]: Done

File: backends/openvino/quantizer/llm_compression.py (hunk 22 above)
Comment (anzr299) [reply backends/openvino/quantizer/llm_compression.py:56]: Done

File: backends/openvino/quantizer/llm_compression.py (hunk 21 above)
Comment (anzr299) [reply backends/openvino/quantizer/llm_compression.py:28]: Done
